### ISSUE TITLE
Packmol code fails to introduce CONECT entries whenever one of the molecules involved is water

### ISSUE THREAD

**Issue by davidlmobley** [choderalab/openmoltools#253]
The `packmol.pack_box` functionality here fails to introduce CONECT entries needed whenever one of the molecules being packed into the box is water, for some reason. 

Apparently packmol never generates CONECT entries, at least per the comment here, but at https://github.com/choderalab/openmoltools/blob/master/openmoltools/packmol.py#L106 there is this remark:
> #Begin hack to introduce bonds for the MISSING CONECT ENTRIES THAT PACKMOL FAILS TO WRITE

For whatever reason, this hack fails to work when one of the molecules is water. Troubleshooting. (CC https://github.com/MobleyLab/SolvationToolkit/issues/78)

**Comment by jchodera** [comment]
Yuck.

What do the CONECT records in your `water.pdb` template files look like?

**Comment by davidlmobley** [comment]
I've traced this a bit further and now am finding that I can take a SINGLE molecule of water in a PDB file without CONECT entries and try to process it via this code and it still fails, even if I have the correct list of bonds. For example, try loading the attached `water.mol2` as the monomer on https://github.com/choderalab/openmoltools/blob/master/openmoltools/packmol.py#L110 and use the attached `packmol.pdb` as your trajectory (https://github.com/choderalab/openmoltools/blob/master/openmoltools/packmol.py#L102). I've manually verified that I get the correct bonds, but no CONECT entries get written when they should.

At this point I'm suspecting an `mdtraj` bug of some sort, @jchodera , though I'd love any insights.



[packmol.pdb.zip](https://github.com/choderalab/openmoltools/files/863727/packmol.pdb.zip)
[water.mol2.zip](https://github.com/choderalab/openmoltools/files/863728/water.mol2.zip)

**Comment by jchodera** [comment]
The [PDB Specification](http://www.wwpdb.org/documentation/file-format-content/format33/sect10.html#CONECT) states
> CONECT records are mandatory for HET groups (excluding water) and for other bonds not specified in the standard residue connectivity table.

That suggests that `CONECT` records should not be required for waters, but should be required for other small molecules. 

OpenMM's `PDBFile` follows this specification closely, and the `Topology.createStandardBonds()` function uses the [definitions in this `residues.xml` file](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/data/residues.xml) to add missing bonds. This file was created from the PDB specifications, and includes the creation of bonds for water residues.

To recap:
* `CONECT` records should be included for non-standard residues that are not water
* `CONECT` records are not required for water or standard amino acid residues

**Comment by jchodera** [comment]
I think this means you'll want to fix the `packmol` tool to produce PDB Specification conformant PDB files if you want to use those in your SolvationToolkit.

**Comment by davidlmobley** [comment]
@jchodera - well, the current issue seems to be that `mdtraj`, for whatever reason (I have yet to dig into the code) is not writing out ANY `CONECT` entries via the mechanism used here when the PDB file in question contains any water molecules, even when the requisite bonds are provided. Presumably I'll have to sort out why -- even if it's not adding them for water, as per the PDB spec, it ought to be for other molecules, whereas I'm finding that if water is present at all, the `CONECT` entries don't get written.

**Comment by jchodera** [comment]
It sounds like it would be useful to file an mdtraj issue.

**Comment by davidlmobley** [comment]
Yup. Sadly have been dealing with logistical stuff (graduate applications, for example) all day, but hopefully I can try and generate some concise code to reproduce and file an issue soon.

**Comment by andrrizzi** [comment]
Hmm, I actually can't reproduce this. Starting from one of the `mol2` in FreeSolv, this script generates CONECT records correctly (i.e. it generates them for the molecule but not for the waters).
```python
import mdtraj as md
from openmoltools.packmol import pack_box

mol_id = 'mobley_5857'
mol_traj = md.load(mol_id + '.mol2')
mol_traj.save_pdb(mol_id + '.pdb')

water_traj = md.load('packmol.pdb')
packed_traj = pack_box([mol_traj, water_traj], [1, 5])

packed_traj.save_pdb('packed.pdb')
```
I'm on Python 3.5, mdtraj 1.7.2, openmoltools 0.7.3, and latest packmol.

**Comment by davidlmobley** [comment]
Hm, well, that's good, then! Maybe I had an issue with my mdtraj version or something. We could mark as closed and I can re-open if I encounter it again.

**Comment by davidlmobley** [comment]
Well, that may not make sense - looks like I'm using `mdtraj` 1.8. Let me try your code, hold on.

**Comment by davidlmobley** [comment]
@andrrizzi - what is your `packmol.pdb` file and where does it come from?

**Comment by andrrizzi** [comment]
It's the one you linked above (`packmol.pdb.zip`). I've now noticed though that `PDBFile` has problems with it because both hydrogen atoms are called `H`. This causes `PDBFile` to raise a warning about duplicate atoms and to ignore the bonds inside water molecules in the topology. Could it be the cause of the error you were experiencing?

**Comment by davidlmobley** [comment]
Could be. I'll have to check that. Thanks for the idea.

**Comment by andrrizzi** [comment]
Sorry, that may be a red herring. I'll do few experiments and get back to you.

**Comment by andrrizzi** [comment]
Ok, it looks like `mdtraj`'s , water is recognized (i.e. bonds are added automatically to the topology when reading a PDB file) only if the residue name is `HOH` and the atom names are `O`, `H1` and `H2` (see [`residues.xml`](https://github.com/mdtraj/mdtraj/blob/master/mdtraj/formats/pdb/data/residues.xml#L506-L509) that define standard residues), so that may actually be the problem.

As far as I can tell, the same behavior is adopted by OpenMM `PDBFile`.

**Comment by davidlmobley** [comment]
Ahh, OK. So that suggests we need to check any waters we provide to `pack_box` before processing to ensure they have correct atom naming.

Should we maybe do this via a SMARTS search to recognize that the molecule is water (or is there an simpler way?) and then renaming the atoms, @andrrizzi ? I can get you how to do the SMARTS match if that sounds simplest.

**Comment by andrrizzi** [comment]
Am I correct assuming that this will introduce an `openeye` dependency in the `pack_box` function? I'd like to avoid that if possible. What about just raising a warning (or an exception if you prefer) if the molecule has 3 atoms, `HOH` or `WAT` residue name and atom names containing two `H` and one `O`? Or something of this kind?

**Comment by davidlmobley** [comment]
Ah, it's true, that WOULD do so. 

On the other hand, it's also extremely annoying to NOT have this work reliably on structures containing water molecules unless those water molecules conform to specific naming conventions. For example, we (in openmoltools) currently assign molecules generated from SMILES random residue names in many situations, so if you generate a water molecule from SMILES you tend to end up with a residue name that is not `HOH` or `WAT`, and the atom names don't necessary conform to `O`, `H1` and `H2`. So, this seems to me to mean that SOMEWHERE we need a mechanism to ensure these get fed into this routine with "correct" names.

Perhaps we should go the opposite route and, here, look for residues which contain only three atoms, one of which is the element O and the other two are the element H, and rename so that `mdraj` will behave appropriately?

**Comment by andrrizzi** [comment]
Good point! I didn't notice your suggestion of renaming the atoms in the first comment. That sounds perfectly reasonable. I'll try to make it work.

**Comment by andrrizzi** [comment]
It turns out I missed that mdtraj defines alternative names for water residue/atom names [here](https://github.com/mdtraj/mdtraj/blob/master/mdtraj/formats/pdb/data/pdbNames.xml#L272).

I've suggested to extend that definition to include `O1` as oxygen (used by OpenEye) and `TP3` as resname. Feel free to chip in if you think the definition should be extended to other naming conventions.

**Comment by andrrizzi** [comment]
Now that I've added the `packmol.standardize_water()` and extended the definition of water on MDTraj to include `O1` I think this can be closed. Feel free to reopen if I forgot something.
